laplace/subnetlaplace.py

```diff
@@ -0,0 +1,129 @@
+    which sums as
+    \\[
+        P = \\sum_{n=1}^N \\nabla^2_\\theta \\log p(\\mathcal{D}_n \\mid \\theta)
+        \\vert_{\\theta_{MAP}} + \\nabla^2_\\theta \\log p(\\theta) \\vert_{\\theta_{MAP}}.
+    \\]
+    The prior is assumed to be Gaussian and therefore we have a simple form for
+    \\(\\nabla^2_\\theta \\log p(\\theta) \\vert_{\\theta_{MAP}} = P_0 \\).
+    In particular, we assume a scalar or diagonal prior precision so that in
+    all cases \\(P_0 = \\textrm{diag}(p_0)\\) and the structure of \\(p_0\\) can be varied.
+
+    The subnetwork Laplace approximation only supports a full, i.e., dense, log likelihood
+    Hessian approximation and hence posterior precision.  Based on the chosen `backend`
+    parameter, the full approximation can be, for example, a generalized Gauss-Newton
+    matrix.  Mathematically, we have \\(P \\in \\mathbb{R}^{P \\times P}\\).
+    See `FullLaplace` and `BaseLaplace` for the full interface.
+
+    References
+    ----------
+    [1] Daxberger, E., Nalisnick, E., Allingham, JU., Antorán, J., Hernández-Lobato, JM.
+    [*Bayesian Deep Learning via Subnetwork Inference*](https://arxiv.org/abs/2010.14689). 
+    ICML 2021.
+
+    Parameters
+    ----------
+    model : torch.nn.Module or `laplace.utils.feature_extractor.FeatureExtractor`
+    likelihood : {'classification', 'regression'}
+        determines the log likelihood Hessian approximation
+    subnetwork_indices : torch.LongTensor
+        indices of the vectorized model parameters
+        (i.e. `torch.nn.utils.parameters_to_vector(model.parameters())`)
+        that define the subnetwork to apply the Laplace approximation over
+    sigma_noise : torch.Tensor or float, default=1
+        observation noise for the regression setting; must be 1 for classification
+    prior_precision : torch.Tensor or float, default=1
+        prior precision of a Gaussian prior (= weight decay);
+        can be scalar, per-layer, or diagonal in the most general case
+    prior_mean : torch.Tensor or float, default=0
+        prior mean of a Gaussian prior, useful for continual learning
+    temperature : float, default=1
+        temperature of the likelihood; lower temperature leads to more
+        concentrated posterior and vice versa.
+    backend : subclasses of `laplace.curvature.CurvatureInterface`
+        backend for access to curvature/Hessian approximations
+    backend_kwargs : dict, default=None
+        arguments passed to the backend on initialization, for example to
+        set the number of MC samples for stochastic approximations.
+    """
+    # key to map to correct subclass of BaseLaplace, (subset of weights, Hessian structure)
+    _key = ('subnetwork', 'full')
+
+    def __init__(self, model, likelihood, subnetwork_indices, sigma_noise=1., prior_precision=1.,
+                 prior_mean=0., temperature=1., backend=BackPackGGN, backend_kwargs=None):
+        self.H = None
```

Comment:
It looks like `_init_H()` will only be called when `fit()` is called with `override=True`, so it would fail for `override=False`, or am I missing something? Maybe it's worth adding a test for that (and maybe also the test for initialization with a large model, but this should work here, since you correctly set `self.H = None` before inheriting anything).

Reply:
Good catch! I believe nothing speaks against calling `_init_H()` at the end of the constructor here, since we now always know `self.n_params_subnet` from the very beginning when passing `subnetwork_indices`, so I'll just add that in to make `fit()` work with `override=True` from the very beginning.

I'll also add the test with the large model!